aioxmpp/ibr/service.py

```diff
@@ -0,0 +1,253 @@
+    account should sent the needed stanza without being authenticated,
+    otherwise it would make no sense.
+
+    .. automethod:: get_client_info
+
+    .. automethod:: change_pass
+
+    .. automethod:: cancel_registration
+
+    .. automethod:: get_registration_fields
+
+    .. automethod:: register
+
+    """
+
+    @asyncio.coroutine
+    def get_client_info(self, client):
+        """
+        A query is sent to the server o obtain the client's data stored at the
+        server.
+
+        :param client: Client from which the query is sent.
+        :type client: :class:`aioxmpp.Client`
+
+        :return: IQ response :attr:`~.IQ.payload` or :data:`None`
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.GET,
+            payload=xso.Query()
+        )
+
+        return (yield from client.send(iq))
+
+    @asyncio.coroutine
+    def change_pass(self,
+                    client,
+                    new_pass,
+                    aux_fields=None):
+        """
+        Change the client password for 'new_pass'.
+
+        :param client: Client who wants to change the password.
+        :type client: :class:`aioxmpp.Client`
+
+        :param new_pass: New password of the client.
+        :type new_pass: :class:`str`
+
+        :param aux_fields: Auxiliary fields in case additional info is needed.
+        :type aux_fields: :class:`dict`
+
+        :return:`None`
+
+        If the query fails for whatever reason, it will raise an exception
+        from inside client.send(iq).
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.SET,
+            payload=xso.Query()
+        )
+
+        iq.payload.username = client.local_jid.localpart
+        iq.payload.password = new_pass
+
+        if aux_fields is not None:
+            for key, value in aux_fields.items():
+                setattr(iq.payload, key, value)
+        yield from client.send(iq)
+
+    @asyncio.coroutine
+    def cancel_registration(self, client):
+        """
+        Cancels the currents client's account with the server.
+
+        :param client: Client who wants to cancel.
+        :type client: :class:`aioxmpp.Client`
+
+        :return:`None`
+
+        Even if the cancelation is succesful, this method will raise an
+        exception due to he account no longer exists for the server, so the
+        client will fail.
+        To continue with the execution, this method should be surrounded by a
+        try/except statement.
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.SET,
+            payload=xso.Query()
+        )
+
+        iq.payload.remove = True
+        yield from client.send(iq)
+
+    @staticmethod
+    @asyncio.coroutine
+    def connect_and_send(iq, loop, metadata, peer_jid, timeout):
```

Comment:
I am not too happy with ``connect_and_send``. This is has extremely high overhead: it does DNS lookups (modulo caching), TCP connect, TLS handshake only to send a single IQ and then abort.

If you look at how aioxmpp core handles establishing a connection (``aioxmpp.node._try_options`` and ``aioxmpp.node.connect_xmlstream``), you will notice that the error handling is also much more complex than what you currently have.

I suggest to, instead of completely re-writing the code which establishes the XML stream yourself, modify ``aioxmpp.node._try_options`` to be able to do what you need. If I’m not mistaken, your requirements are:

1. Get an XMLStream
2. Without authentication (pre-SASL)

Achieving this can be as simple as adding an ``if`` to ``_try_options`` which checks if any SASL providers are given in the security context (``metadata``). If the list is empty, simply skip sasl negotiation. You will have to modify ``node.py`` and write a few simple tests for ``_try_options``, but that’s okay. We are adding a new feature to aioxmpp core after all (which is: a way to handle pre-authentication things).


You can then replace ``connect_and_send`` with ``connect_xmlstream`` with a metadata which has an empty list of ``sasl_providers``. The usage of the IBR API would then look roughly like this:

1. ``_, xmlstream, features = connect_xmlstream()``
2. check if IBR offered in ``features``, otherwise fail early
3. use ``aioxmpp.ibr.get_registration_fields(xmlstream)`` to obtain the fields required
4. fill out the fields
5. invoke ``aioxmpp.ibr.register(xmlstream, fields)``